Example of validation (#169)
arrow-kt/arrow-website: content/docs/learn/typed-errors/validation.md

---
id: validation
title: Validation
description: Worked out example of validation.
sidebar_position: 4
---

# Validation

<!--- TEST_NAME ValidationTest -->

<!--- INCLUDE .*
import arrow.core.left
import arrow.core.right
import arrow.core.Either
import arrow.core.NonEmptyList
import arrow.core.recover
import arrow.core.raise.*
-->

This tutorial shows a concrete example of using typed errors to implement
domain validation. In particular, we begin with the following domain:

```kotlin
data class Author(val name: String)

> data class Book(val title: String, val authors: List<Author>)

Reviewer: Should `authors` here be `NonEmptyList`? 🤔

```
<!--- KNIT example-validation-01.kt -->

over which we want to implement the following rules:

1. The given title should not be empty,
2. The list of authors should not be empty,
3. None of the author names should be empty.

We want to accumulate as many error as possible.

## Smart constructors

The `Author` class exposes its constructor, so we cannot prevent from creating
wrong values. One could introduce a `require` in the constructor, but we prefer

> to use the typed error mechanism instead. The usual pattern in this case is to
> _hide_ the constructor, and provide a _smart constructor_ by adding an `invoke`
> function in the companion.

Reviewer: ```suggestion
to use the typed error mechanism instead. A common pattern in this case is to
_hide_ the constructor, and provide a _smart constructor_ by adding an `invoke`
operator in the companion.
```

Should we also discuss the trade-offs somewhere? It's been a topic that has been discussed many times, and everytime we show this in a workshop we also get questions around this topic.

To summarise (if memory serves me right):

##### `operator fun invoke`
 - Pro: allows you to _mimic_ a real constructor call, **and** allow you to make the `constructor` private.
 - Pro: Doesn't require additional import
 - Con: Click-through on `Author` brings you `companion object`, you need to click `(..)` braces to find definition site

##### named companion function
 - Pro: allow you to make the `constructor` private.
 - Con: requires additional import
 - Con: Doesn't mimic real constructor call
 
#### Top-level `fun Author`
 - Pro: allows you to _mimic_ a real constructor call, **but** doesn't allow you to make the `constructor` private.
 - Con: doesn't allow the `constructor` to be private
 - Con: Requires additional import

cc\\ @clovis-ai

Reviewer 2: Ahah, I love that you cc'd me here. Sorry for being late, but another important consideration for me is how it reads for beginners: if I share the link to a Dokka page of an `operator fun` function in a companion object to someone who just started learning Kotlin, they have no idea how to call it. Both other solutions are obvious (and IDEA will take care of the imports).

a hierarchy. Note that `EmptyAuthor` is a different type than before,
since we want to accommodate the index of the author.

<!--- INCLUDE
data class Author(val name: String)
-->

```kotlin
sealed interface BookValidationError
object EmptyTitle: BookValidationError
object NoAuthors: BookValidationError
data class EmptyAuthor(val index: Int): BookValidationError

data class Book private constructor(val title: String, val authors: List<Author>) {
  companion object {
    operator fun invoke(
      title: String, authors: List<String>
    ): Either<BookValidationError, Book> = TODO()
  }
}
```
<!--- KNIT example-validation-04.kt -->

Let's forget about validating each author for a moment, and just implement the
emptiness checks for the title and the authors list.

<!--- INCLUDE
sealed interface BookValidationError
object EmptyTitle: BookValidationError
object NoAuthors: BookValidationError
data class EmptyAuthor(val index: Int): BookValidationError

data class Author(val name: String)
-->

```kotlin
data class Book private constructor(val title: String, val authors: List<Author>) {
  companion object {
    operator fun invoke(

>       title: String, authors: List<String>

Reviewer: ```suggestion
      title: String, authors: Iterable<String>
```

This could be `Iterable`, whilst `Book` contains a `NonEmptyList` property.

object EmptyTitle: BookValidationError
object NoAuthors: BookValidationError
data class EmptyAuthor(val index: Int): BookValidationError

data class Author(val name: String)
-->

```kotlin
data class Book private constructor(val title: String, val authors: List<Author>) {
  companion object {
    operator fun invoke(
      title: String, authors: List<String>
    ): Either<NonEmptyList<BookValidationError>, Book> = either {
      zipOrAccumulate(
        { ensure(title.isNotEmpty()) { EmptyTitle } },
        { ensure(authors.isNotEmpty()) { NoAuthors } }
      ) { _, _ ->
        Book(title, TODO())
      }
    }
  }
}
```
<!--- KNIT example-validation-06.kt -->

The result of each of the arguments of `zipOrAccumulate` are available in the
trailing lambda. In this case we are not using them: we have the `title`
already available, and for the list of authors we still need to perform the
conversion from `String` to `Author`.

:::tip Arguments to zipOrAccumulate

All but the last argument to `zipOrAccumulate` represent the different
validations we want to run aggregating their output. Those arguments are
often _blocks_ wrapped in `{` curly braces `}`, which is a bit unusual for

> most Kotliners.

Reviewer: > which is a bit unusual for most Kotliners.
 
🤔 Is it? I was hoping it wouldn't be unusual 😅

Author: I got from the workshop that most people are not used to it.

Reviewer: Damn, that's too bad. Should we include a section on `Either.zipOrAccumulate` which works _value based_?

:::

## Validating a list

The next step is turning the given `authors`, which is a list of `String`s, into
a list of `Author`s. We need to run the smart constructor, thus, but at the same
time we should accumulate any potential problems. Since this is related to
author checks, we'll include this as part of the second validation.

<!--- INCLUDE
sealed interface BookValidationError
object EmptyTitle: BookValidationError
object NoAuthors: BookValidationError
data class EmptyAuthor(val index: Int): BookValidationError

object EmptyAuthorName

data class Author private constructor(val name: String) {
  companion object {
    operator fun invoke(name: String): Either<EmptyAuthorName, Author> = either {
      ensure(name.isNotEmpty()) { EmptyAuthorName }
      Author(name)
    }
  }
}
-->

```kotlin
data class Book private constructor(val title: String, val authors: List<Author>) {
  companion object {
    operator fun invoke(
      title: String, authors: List<String>
    ): Either<NonEmptyList<BookValidationError>, Book> = either {
      zipOrAccumulate(
        { ensure(title.isNotEmpty()) { EmptyTitle } },
        { 
          ensure(authors.isNotEmpty()) { NoAuthors }

>           authors.withIndex().mapOrAccumulate {
>             Author(it.value)
>               .recover { _ -> raise(EmptyAuthor(it.index)) }
>               .bind()
>           }

Reviewer: I think this code is incorrect, and is missing the `bind` call. Otherwise the syntax is `mapOrAccumulate(authors.withIndex()) { ... }`

```suggestion
          authors.withIndex().mapOrAccumulate {
            Author(it.value)
              .recover { _ -> raise(EmptyAuthor(it.index)) }
              .bind()
          }.bind()
```

        }
      ) { _, validatedAuthors ->
        Book(title, validatedAuthors)
      }
    }
  }
}
```
<!--- KNIT example-validation-07.kt -->

This additional check is quite complex, so let's unravel it step by step:
- We use `authors.withIndex()` to create an iterable containing the values
  in conjunction with the index they live in, this is necessary to create
  the right `EmptyAuthor` error value.
- With `mapOrAccumulate` we state that we want to perform some validation over
  a collection of elements, accumulating each possible error.
- The call to `Author(it.value)` return an `Either` with the wrong error type
  (`EmptyAuthorName` instead of `EmptyAuthor`). To transform this value we
  use the `recover` extension function.

:::tip Recover vs. Map Left

Another possibility would be to use `mapLeft { EmptyAuthor(it.index) }`.
The difference between `recover` and `mapLeft` is that the latter only transforms
the error value, whereas in the former you can use any typed error computation.

:::

- Finally, we use `.bind()` to embed the `Either` into the computation block.

>   Essentially, every time you are using a value of type `Either` inside an
>   `either` block, such a call is required.

Reviewer: ```suggestion
  Essentially, every time you are using a value of type `Either` inside an
  `either` or `Raise` block, such a call is required.
```

object NoAuthors: BookValidationError
data class EmptyAuthor(val index: Int): BookValidationError

object EmptyAuthorName

data class Author private constructor(val name: String) {
  companion object {
    operator fun invoke(name: String): Either<EmptyAuthorName, Author> = either {
      ensure(name.isNotEmpty()) { EmptyAuthorName }
      Author(name)
    }
  }
}

data class Book private constructor(val title: String, val authors: List<Author>) {
  companion object {
    operator fun invoke(
      title: String, authors: List<String>
    ): Either<NonEmptyList<BookValidationError>, Book> = either {
      zipOrAccumulate(
        { ensure(title.isNotEmpty()) { EmptyTitle } },
        { 
          ensure(authors.isNotEmpty()) { NoAuthors }
          authors.withIndex().mapOrAccumulate { a ->
            Author(a.value).mapLeft { EmptyAuthor(a.index) }
          }.bindAll()
        }
      ) { _, validatedAuthors ->
        Book(title, validatedAuthors)
      }
    }
  }
}
-->
<!--- KNIT example-validation-08.kt -->

:::info Bind All

Another way to write the code above is creating a list of `Either` using
`mapOrAccumulate`, and then using `.bindAll()` at the very end.

```

> authors.withIndex().mapOrAccumulate { nameAndIx ->
>   Author(nameAndIx.value).mapLeft { EmptyAuthor(nameAndIx.index) }
> }.bindAll()

Reviewer: ```suggestion
authors.withIndex().map { nameAndIx ->
  Author(nameAndIx.value).mapLeft { EmptyAuthor(nameAndIx.index) }
}.bindAll()
```

object NoAuthors: BookValidationError
data class EmptyAuthor(val index: Int): BookValidationError

object EmptyAuthorName

data class Author private constructor(val name: String) {
  companion object {
    operator fun invoke(name: String): Either<EmptyAuthorName, Author> = either {
      ensure(name.isNotEmpty()) { EmptyAuthorName }
      Author(name)
    }
  }
}

data class Book private constructor(val title: String, val authors: NonEmptyList<Author>) {
  companion object {
    operator fun invoke(
      title: String, authors: Iterable<String>
    ): Either<NonEmptyList<BookValidationError>, Book> = either {
      zipOrAccumulate(
        { ensure(title.isNotEmpty()) { EmptyTitle } },
        { 
          val validatedAuthors = authors.withIndex().map { a ->
            Author(a.value).mapLeft { EmptyAuthor(a.index) }
          }.bindAll()
          ensureNotNull(validatedAuthors.toNonEmptyListOrNull()) { NoAuthors }
        }
      ) { _, validatedAuthors ->
        Book(title, validatedAuthors)
      }
    }
  }
}
-->
<!--- KNIT example-validation-08.kt -->

:::info Bind All

Another way to write the code above is creating a list of `Either` using

> `mapOrAccumulate`, and then using `.bindAll()` at the very end.

Reviewer 3: You mention `mapOrAccumulate` in the text. However, the code example uses just `map`. Both cases compile because `zipOrAccumulate` already provides the `RaiseAccumulate` context, but I think we should explain it or fix the text.

Author: I've updated and clarified the text, does it look better now?

Reviewer 3: Looks perfect now, thanks @serras
